File: src/delta_exchange_mcp/connection_app.py
"""MCP App that opens the external Manage Connection page."""

from mcp.server.apps import APP_MIME_TYPE

VIEW_URI = "ui://delta-exchange/manage-connection.html"
VIEW_MIME = APP_MIME_TYPE

VIEW_HTML = """<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title>Manage Delta Exchange connection</title>
<style>
  body { font-family: system-ui, sans-serif; margin: 0; padding: 1em; }
  button { font: inherit; min-height: 2.75em; padding: .5em 1em; }
  p { margin: 0 0 .75em; }
</style>
</head>
<body>
<p id="status">Opening the secure connection page…</p>
<button id="open" type="button" hidden>Open connection page</button>
<script>
(() => {
  "use strict";
  const status = document.getElementById("status");
  const button = document.getElementById("open");
  const pending = new Map();
  const appsVersion = "2026-01-26";
  let nextId = 1;
  let manageUrl = "";
  let initialized = false;
  let canOpenLinks = false;

  function post(message) { window.parent.postMessage(message, "*"); }
  function request(method, params) {
    const id = nextId++;
    post({ jsonrpc: "2.0", id, method, params: params || {} });
    return new Promise((resolve, reject) => pending.set(id, { resolve, reject }));
  }
  async function open() {
    if (!manageUrl || !initialized) return;
    if (!canOpenLinks) {
      status.textContent = "Open the connection link in the tool response.";
      button.hidden = true;
      return;
    }
    try {
      const result = await request("ui/open-link", { url: manageUrl });
      if (result && result.isError) throw new Error("Open link failed");
      status.textContent = "Connection page opened.";
    } catch {
      status.textContent = "This client could not open the connection page.";
      button.hidden = false;
    }
  }
  function capture(result) {
    const ui = result && result._meta && result._meta.ui;
    if (!ui || !ui.manageUrl) return;
    manageUrl = ui.manageUrl;
    button.hidden = false;
    open();
  }

  window.addEventListener("message", event => {
    if (event.source !== window.parent) return;
    const message = event.data;
    if (!message || message.jsonrpc !== "2.0") return;
    if (message.id !== undefined && message.method === undefined) {
      const callback = pending.get(message.id);
      if (!callback) return;
      pending.delete(message.id);
      if (message.error) callback.reject(new Error(message.error.message));
      else callback.resolve(message.result);
      return;
    }
    if (message.method === "ui/notifications/tool-result") capture(message.params);
  });

  button.addEventListener("click", open);
  request("ui/initialize", {
    appCapabilities: {},
    appInfo: { name: "Delta Exchange connection", version: "1" },
    protocolVersion: appsVersion,
  }).then(result => {
    if (!result || result.protocolVersion !== appsVersion) {
      throw new Error("Unsupported MCP Apps version");
    }
    initialized = true;
    canOpenLinks = Boolean(result.hostCapabilities && result.hostCapabilities.openLinks);
    post({ jsonrpc: "2.0", method: "ui/notifications/initialized" });
    if (result.toolResult) capture(result.toolResult);
    else open();
  }).catch(() => {
    status.textContent = "This client could not initialize the connection view.";
  });
})();
</script>
</body>
</html>
"""

File: tests/connection_app.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import vm from "node:vm";

const source = readFileSync(0, "utf8");
const version = "2026-01-26";
const url = "http://127.0.0.1:12345/disposable-test-path";
const toolResult = { _meta: { ui: { manageUrl: url } } };

function view() {
  const messages = [];
  const listeners = new Map();
  const elements = new Map(["status", "open"].map(id => [id, {
    textContent: "",
    hidden: true,
    addEventListener: (event, callback) => listeners.set(`${id}:${event}`, callback),
  }]));
  const parent = { postMessage: message => messages.push(message) };
  const window = { parent, addEventListener: (event, callback) => listeners.set(event, callback) };
  const document = { getElementById: id => elements.get(id) };
  vm.runInNewContext(source, { window, document });
  const receive = data => listeners.get("message")({ source: parent, data: { jsonrpc: "2.0", ...data } });
  const initialize = result => receive({ id: messages[0].id, result });
  return { messages, elements, listeners, receive, initialize };
}

const flush = () => new Promise(resolve => setImmediate(resolve));

for (const failure of [false, "result", "rpc"]) {
  const app = view();
  assert.equal(app.messages[0].params.protocolVersion, version);
  app.receive({ method: "ui/notifications/tool-result", params: toolResult });
  assert.equal(app.messages.length, 1, "do not open before negotiation");
  app.initialize({ protocolVersion: version, hostCapabilities: { openLinks: {} } });
  await flush();
  assert.equal(app.messages[1].method, "ui/notifications/initialized");
  const open = app.messages[2];
  assert.equal(open.method, "ui/open-link");
  assert.equal(open.params.url, url);
  app.receive(failure === "rpc"
    ? { id: open.id, error: { code: -32603, message: "cannot open" } }
    : { id: open.id, result: { isError: failure === "result" } });
  await flush();
  assert.equal(app.elements.get("status").textContent, failure
    ? "This client could not open the connection page."
    : "Connection page opened.");
  if (failure) {
    assert.equal(app.elements.get("open").hidden, false);
    app.listeners.get("open:click")();
    assert.equal(app.messages.at(-1).method, "ui/open-link");
  }
}

const unsupported = view();
unsupported.initialize({ protocolVersion: "2026-07-28", hostCapabilities: { openLinks: {} }, toolResult });
await flush();
assert.equal(unsupported.messages.length, 1);
assert.equal(unsupported.elements.get("status").textContent, "This client could not initialize the connection view.");

const noLinks = view();
noLinks.initialize({ protocolVersion: version, hostCapabilities: {}, toolResult });
await flush();
assert.equal(noLinks.messages.length, 2);
assert.equal(noLinks.elements.get("open").hidden, true);
assert.equal(noLinks.elements.get("status").textContent, "Open the connection link in the tool response.");
